external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -118,6 +123,7 @@ class DirectKafkaInputDStream[
       context.sparkContext, kafkaParams, currentOffsets, untilOffsets, messageHandler)
 
     currentOffsets = untilOffsets.map(kv => kv._1 -> kv._2.offset)
+    offsetMap += ((validTime, currentOffsets))
```

Comment:
if it's not checkpointed, is rememberDuration guaranteed to be high enough that the rdd hasn't already been cleared out by the time the batch completion listener runs?

If that's all ok, sure, seems much better to me than keeping the redundant info around.